book/src/dev/rfcs/0012-value-pools.md

```diff
@@ -0,0 +1,498 @@
+
+- Method location is at `zebra-chain/src/transparent.rs`.
+
+```rust
+impl Output {
+    fn value_balance(&self) -> ValueBalance<NegativeAllowed> {
+
+    }
+}
+```
+
+#### Create a method in `JoinSplitData` that returns `ValueBalance<NegativeAllowed>`
+
+- Method location is at `zebra-chain/src/transaction/joinsplit.rs`
+
+```rust
+pub fn value_balance(&self) -> ValueBalance<NegativeAllowed> {
+
+}
+```
+
+#### Create a method in `sapling::ShieldedData` that returns `ValueBalance<NegativeAllowed>`
+
+- Method location is at `zebra-chain/src/transaction/sapling/shielded_data.rs`
+
+```rust
+pub fn value_balance(&self) -> ValueBalance<NegativeAllowed> {
+
+}
+```
+
+#### Create a method in `orchard::ShieldedData` that returns `ValueBalance<NegativeAllowed>`
+
+- Method location is at `zebra-chain/src/transaction/orchard/shielded_data.rs`
+
+```rust
+pub fn value_balance(&self) -> ValueBalance<NegativeAllowed> {
+
+}
+```
+
+#### Create the `Transaction` method
+
+- Method location: `zebra-chain/src/transaction.rs`
+
+```rust
+/// utxos must contain the utxos of every input in the transaction,
+/// including UTXOs created by earlier transactions in this block.
+pub fn value_balance(&self, utxos: &HashMap<transparent::OutPoint, Utxo>) -> Result<ValueBalance<NegativeAllowed>, Err> {
+
+}
+```
+
+### Create a method in `Block` that returns `ValueBalance<NegativeAllowed>` for the block
+
+- Method location is at `zebra-chain/src/block.rs`.
+- Method will make use of `Transaction::value_balance` method created before.
+
+```rust
+/// utxos must contain the utxos of every input in the transaction,
+/// including UTXOs created by a transaction in this block,
+/// then spent by a later transaction that's also in this block.
+pub fn value_balance(&self, utxos: &HashMap<transparent::OutPoint, Utxo>) -> ValueBalance<NegativeAllowed> {
+    self.transactions()
+        .map(Transaction::value_balance)
+        .sum()
+        .expect("Each block should have at least one coinbase transaction")
+}
+```
+
+### Pass the value balance for this block from the consensus into the state
+
+- Add a new field into `PreparedBlock` located at `zebra-state/src/request.rs`, this is the `NonFinalized` section of the state.
+
+```rust
+pub struct PreparedBlock {
+    ..
+    /// The value balances for each pool for this block.
+    pub block_value_balance: ValuePool<NegativeAllowed>,
+}
+```
+- In `zebra-consensus/src/block.rs` pass the value balance to the zebra-state:
+
+```rust
+let block_value_balance = block.value_balance();
+let prepared_block = zs::PreparedBlock {
+    ..
+    block_value_balance,
+};
+```
+
+### Add a value pool into the state `Chain` struct
+
+- This is the value pool for the non finalized part of the blockchain.
+- Location of the `Chain` structure where the pool field will be added: `zebra-state/src/service/non_finalized_state/chain.rs`
+
+```rust
+pub struct Chain {
+    ..
+    value_pool: ValueBalance<NonNegative>,
```

Comment:
```suggestion
    /// The chain value pool balance at the tip of this chain.
    value_pool: ValueBalance<NonNegative>,
```

Reply:
Thanks, added this at https://github.com/ZcashFoundation/zebra/pull/2430/commits/3af230cea52cad2b005034dfcac9cda9c0a16ba0